### ISSUE TITLE
kube-vip services drops the VIP, doesn't retake it without manual restart

### ISSUE THREAD

**Issue by mdbooth** [kube-vip/kube-vip#475]
TL;DR

This isn't really a bug report; it's more of a feature request. The structure of the services controller is nothing like I have ever seen. It's extremely hard to reason about. I suspect it contains a lot of races, more than one of which are likely causing the issues I hit daily. I was going to try to find and fix the bugs I'm hitting, but I'm left thinking I'd rather re-architect the control loop.

The control plane VIP has been *solid*, though 👍

**Describe the bug**
I run kube-vip in a small, unreliable cluster. The hardware's janky. The network's janky. Everything's janky. Something somewhere is failing most of the time.

The cluster is deployed by kubeadm. I deploy kube-vip as a static pod on the control plane. YAML below.

In this environment kube-vip LB services frequently get wedged in a state where a node has dropped the service VIP but no other node has taken it. The lease has expired, so presumably the other nodes weren't watching it. I run most most of my services with `externalTrafficPolicy: Local`, so scheduling may also be a factor.

I see at least 4 different failure modes in the kube-vip logs:

1. Nothing. Everything's fine, apparently, but the node doesn't have the VIP.
2. Screens of kube-apiserver connection errors from a variety of controller sources (retrywatcher, leaderelection).
3. Screens of *client-side* rate throttling because the client is looping with no backoff.
4. Something happened, the node lost the service election and dropped the VIP, but never took it back.

These may or may not have common causes. As mentioned above I suspect a variety of causes.

In all cases, manually killing kube-vip will cause it to recover *immediately*. This is important because it suggests that whatever condition caused kube-vip to drop the VIP in the first place is no longer present, but kube-vip wasn't able to recover in its current state.

The code structure of the services handling is unusual. I've only spent an hour or so browsing, so apologies if I've missed something very important. It seems to involve setting separate watches on individual services. Apart from being harder to reason about concurrency, I suspect this has also has a higher overhead on the apiserver than a single watch on services. Is there any reason you've done it this way rather than something more like, e.g. the controller runtime approach? I guess the complexity here is the multiple leader elections? I suspect we're missing numerous ordering edge cases which are fairly easy to hit in practise once you start introducing the occasional failure and a bit of latency. Having a single control loop which is able to reconstruct state on each iteration should make that a lot more failure tolerant in practise.

Anyway, this is something I'm potentially willing and able to work on. Is that something you're interested in? If so, any insight into how we got here?

**Environment (please complete the following information):**
 - OS/Distro: Fedora CoreOS
 - Kubernetes Version: 1.25
 - Kube-vip Version: 0.5.5

**`Kube-vip.yaml`:**
```yaml
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: null
  name: kube-vip
  namespace: kube-system
spec:
  containers:
  - args:
    - manager
    env:
    - name: vip_arp
      value: "true"
    - name: port
      value: "6443"
    - name: vip_interface
      value: eno1
    - name: vip_cidr
      value: "32"
    - name: cp_enable
      value: "true"
    - name: cp_namespace
      value: kube-system
    - name: vip_ddns
      value: "false"
    - name: svc_enable
      value: "true"
    - name: svc_election
      value: "true"
    - name: vip_leaderelection
      value: "true"
    - name: vip_leaseduration
      value: "5"
    - name: vip_renewdeadline
      value: "3"
    - name: vip_retryperiod
      value: "1"
    - name: lb_enable
      value: "true"
    - name: lb_port
      value: "6443"
    - name: lb_fwdmethod
      value: local
    - name: address
      value: api.cluster.heisenbug.com
    - name: prometheus_server
      value: :2112
    image: ghcr.io/kube-vip/kube-vip:v0.5.5
    imagePullPolicy: Always
    name: kube-vip
    resources: {}
    securityContext:
      capabilities:
        add:
        - NET_ADMIN
        - NET_RAW
    volumeMounts:
    - mountPath: /etc/kubernetes/admin.conf
      name: kubeconfig
  hostAliases:
  - hostnames:
    - kubernetes
    ip: 127.0.0.1
  hostNetwork: true
  volumes:
  - hostPath:
      path: /etc/kubernetes/admin.conf
    name: kubeconfig
```

**Comment by theodiem** [comment]
So I have a not so janky cluster, but still it is a small deployment, which I have some RPI 4's as master nodes to have enough resiliency without hiking electricity bill, and I came across the same issue, that kube-vip can time-out on leaderelection, releases the VIP and never try to get it back, which is an issue when traffic policy is set to `Local` and only one instance of the service is running.

Example from logs:
```
time="2022-11-06T14:05:23Z" level=info msg="DHCP VIP [0.0.0.0] for [app-docscan/app-docscan-smb] "
time="2022-11-06T14:05:23Z" level=info msg="adding VIP [0.0.0.0] for [app-docscan/app-docscan-smb] "
time="2022-11-06T14:05:23Z" level=info msg="Starting advertising address [10.192.1.169] with kube-vip"
time="2022-11-06T14:05:23Z" level=info msg="Started Load Balancer and Virtual IP"
...
E1106 14:11:57.574890       1 leaderelection.go:367] Failed to update lock: Put "https://172.30.0.1:443/apis/coordination.k8s.io/v1/namespaces/app-docscan/leases/kubevip-app-docscan-smb": context deadline exceeded
I1106 14:11:57.574927       1 leaderelection.go:283] failed to renew lease app-docscan/kubevip-app-docscan-smb: timed out waiting for the condition
time="2022-11-06T14:11:57Z" level=info msg="service [app-docscan-smb] leader lost: cdco-nuc-01"
time="2022-11-06T14:11:57Z" level=info msg="[LOADBALANCER] Stopping load balancers"
time="2022-11-06T14:11:57Z" level=info msg="[VIP] Releasing the Virtual IP [10.192.1.169]"
time="2022-11-06T14:11:57Z" level=info msg="release, lease: &{Offer:DHCPv4(xid=0xb2bf8161 hwaddr=00:0d:c2:c7:5b:d9 msg_type=REQUEST, your_ip=0.0.0.0, server_ip=0.0.0.0) ACK:DHCPv4(xid=0xb2bf8161 hwaddr=00:0d:c2:c7:5b:d9 msg_type=ACK, your_ip=10.192.1.169, server_ip=10.192.0.1) CreationTime:2022-11-06 14:05:23.247217242 +0000 UTC m=+0.275224002}"
time="2022-11-06T14:11:57Z" level=info msg="Removed [2ddebb5b-314d-4a4e-a483-547323acf9f2] from manager, [2] advertised services remain"
```
And nothing after this. If I kill the kube-vip pod on that node, it picks up the VIP again immediately.

Although I understand that Raspberry PI's aren't that powerful, I think it would be good for kube-vip to attempt to re-acquire the lock (maybe a configuration key with the retry delay? - Not sure if the election timeout is configurable).

**Comment by thebsdbox** [comment]
I've made a number of changes to the `main` release of kube-vip (which is the nightly, this fixes an issue with some hanging services).

**Comment by thebsdbox** [comment]
The reason for leaderElection per service (which is optional) is to effectively allow some level of load-balancing of services across worker nodes, without it then the elected leader will host ALL services.

**Comment by theodiem** [comment]
> I've made a number of changes to the main release of kube-vip (which is the nightly, this fixes an issue with some hanging services).

@thebsdbox just FYI:

I've been running ghcr.io/kube-vip/kube-vip-iptables@sha256:da45855479af850fdbcf95b56b516d0f12f6e04753b56b810517ec40f4e07be5 for ~5d and services are getting dropped and don't ever come back (until a restart) when the lock times out:

```
E1116 19:47:09.803076       1 leaderelection.go:367] Failed to update lock: Put "https://172.30.0.1:443/apis/coordination.k8s.io/v1/namespaces/app-docscan/leases/kubevip-app-docscan-smb": context deadline exceeded
I1116 19:47:09.803149       1 leaderelection.go:283] failed to renew lease app-docscan/kubevip-app-docscan-smb: timed out waiting for the condition
E1116 19:47:09.817287       1 leaderelection.go:306] Failed to release lock: Operation cannot be fulfilled on leases.coordination.k8s.io "kubevip-app-docscan-smb": the object has been modified; please apply your changes to the latest version and try again
time="2022-11-16T19:47:09Z" level=info msg="[services election] service [app-docscan-smb] leader lost: [cdco-nuc-01]"
```

**Comment by linsite** [comment]
Maybe related with https://github.com/kubernetes/client-go/issues/1155.

**Comment by ChriZ982** [comment]
@thebsdbox this still happens in v0.5.10 for me once a day

Is there any workaround or is the `servicesElection` feature broken and therefore having `externalTrafficPolicy: Local` not possible with kube-vip at the moment?

**Comment by si458** [comment]
i think the issue is still around as i have this issue currently,
i use `servicesElection` with `externalTrafficPolicy: Local` and BGP and it works great, 
UNTIL the node has problems with say high IO usage, then the leader is lost so it releases the IP then a leader is elected again
but then the IP is never assigned to the node again or advertised by BGP until i manually restart the kube-vip daemonset
im using kube-vip 0.6.0 with talos

**Comment by dburkland** [comment]
I'm seeing the exact thing on my new cluster

**Comment by jamestutton** [comment]
I'm seeing the exact thing been driving me mad as until understand it seems semi erratic failure

**Comment by theodiem** [comment]
Hello @Cellebyte ,
Updating from 0.6.3 to 0.7.0, the control panel feature doesn't start anymore. Console output is:
```
....
time="2024-02-08T08:32:22Z" level=info msg="Starting IPVS LoadBalancer"
time="2024-02-08T08:32:22Z" level=info msg="IPVS Loadbalancer enabled for 1.2.1"
time="2024-02-08T08:32:22Z" level=info msg="Gratuitous Arp broadcast will repeat every 3 seconds for [10.192.1.166]"
time="2024-02-08T08:32:22Z" level=info msg="Kube-Vip is watching nodes for control-plane labels"
time="2024-02-08T08:32:22Z" level=error msg="Error querying backends file does not exist"
time="2024-02-08T08:32:22Z" level=error msg="Unable to create an IPVS service, ensure IPVS kernel modules are loaded"
time="2024-02-08T08:32:22Z" level=fatal msg="IPVS service error: netlink receive: invalid argument"
```

(Using ARP mode)
Have any module requirement changed?
```
# ls ./6.1.38-intel-pk-standard/kernel/net/netfilter/ipvs
ip_vs.ko     ip_vs_ftp.ko    ip_vs_lc.ko  ip_vs_ovf.ko     ip_vs_sed.ko   ip_vs_wlc.ko
ip_vs_dh.ko  ip_vs_lblc.ko   ip_vs_mh.ko  ip_vs_pe_sip.ko  ip_vs_sh.ko    ip_vs_wrr.ko
ip_vs_fo.ko  ip_vs_lblcr.ko  ip_vs_nq.ko  ip_vs_rr.ko      ip_vs_twos.ko
# lsmod | grep vs
ip_vs_sh               16384  0
ip_vs_wrr              16384  0
ip_vs_rr               16384  88
ip_vs                 155648  94 ip_vs_rr,ip_vs_sh,ip_vs_wrr
nf_conntrack          118784  5 xt_conntrack,nf_nat,nf_conntrack_netlink,xt_MASQUERADE,ip_vs
nf_defrag_ipv6         20480  2 nf_conntrack,ip_vs
```

For now rolling back to 0.6.3, but I can test again anytime as the cluster is not production.

**Comment by Cellebyte** [comment]
@theodiem I was more referring to the ticket content which mentionend LoadBalancer for the controlplane mode there were no fixes introduced.

**Comment by theodiem** [comment]
As mentioned earlier in the ticket, I also do suffer from the issue that kube-vip stops responding for a LoadBalancer IP address. I believe my issue with modules is that I don't have ipvs ipv6 modules, which is now causing kube-vip for the controlplane to crash with the mentioned message.
I will try enabling ipv6 modules but it might take a while for me to build it since its a custom distro (yocto based).

**Comment by Cellebyte** [comment]
@mdbooth could you test the latest release?

**Comment by thebsdbox** [comment]
The latest release`v0.8.8` has a fix that ensures that load balancers will restart in the event of an API failure.
